### PR TITLE
add timer fallbacker unit test

### PR DESCRIPTION
**Pull request by EvanLjp**
add timer fallbacker unit test

### CODE REVIEW

**Review by wu-sheng — APPROVED** [review]

File: plugins/fallbacker/timer/timer_fallbacker.go [hunk 1]
```diff
@@ -25,16 +25,18 @@ import (
...
-// and the time for each retry is expanded by 2 times until the number of retries reaches the maximum.
+const Name = "timer-fallbacker"
+
+// Fallbacker is a timer fallbacker when forward fails.
 type Fallbacker struct {
 	config.CommonFields
-	maxTimes      int `mapstructure:"max_times"`
-	latencyFactor int `mapstructure:"latency_factor"`
+	MaxTimes       int `mapstructure:"max_times"`
+	LatencyFactor  int `mapstructure:"latency_factor"`
```
Comment (kezhenxu94) [inline plugins/fallbacker/timer/timer_fallbacker.go:34]: s/LatencyFactor/ExponentialBackoff

File: plugins/fallbacker/timer/timer_fallbacker.go [hunk 2]
```diff
@@ -25,16 +25,18 @@ import (
...
-// Fallbacker is a timer fallbacker when forward fails. `latencyFactor` is the standard retry duration,
-// and the time for each retry is expanded by 2 times until the number of retries reaches the maximum.
+const Name = "timer-fallbacker"
+
+// Fallbacker is a timer fallbacker when forward fails.
 type Fallbacker struct {
 	config.CommonFields
-	maxTimes      int `mapstructure:"max_times"`
-	latencyFactor int `mapstructure:"latency_factor"`
+	MaxTimes       int `mapstructure:"max_times"`
```
Comment (kezhenxu94) [inline plugins/fallbacker/timer/timer_fallbacker.go:33]: s/MaxTimes/MaxAttempts

File: plugins/fallbacker/timer/timer_fallbacker.go [hunk 3]
```diff
@@ -43,17 +45,25 @@ func (t *Fallbacker) Description() string {
...
+# The max retry latency time.(Time unit is millisecond.)
+max_latency_time: 5000
 `
 }
 
 func (t *Fallbacker) FallBack(batch event.BatchEvents, forward api.ForwardFunc) bool {
-	currentLatency := t.latencyFactor
-	for i := 1; i < t.maxTimes; i++ {
+	currentLatency := t.LatencyFactor
+	for i := 1; i < t.MaxTimes; i++ {
```
Comment (kezhenxu94) [inline plugins/fallbacker/timer/timer_fallbacker.go:60]:
```suggestion
	for i := 1; i <= t.MaxTimes; i++ {
```

File: plugins/fallbacker/timer/timer_fallbacker.go [hunk 4]
```diff
@@ -43,17 +45,25 @@ func (t *Fallbacker) Description() string {
...
 		time.Sleep(time.Duration(currentLatency) * time.Millisecond)
 		if err := forward(batch); err != nil {
 			currentLatency *= 2
+			if currentLatency > t.MaxLatencyTIme {
```
Comment (kezhenxu94) [inline plugins/fallbacker/timer/timer_fallbacker.go:64]:
```suggestion
			if currentLatency >= t.MaxLatencyTIme {
```

File: plugins/fallbacker/timer/timer_fallbacker.go [hunk 5]
```diff
@@ -43,17 +45,25 @@ func (t *Fallbacker) Description() string {
...
+				currentLatency = t.MaxLatencyTIme
```
Comment (kezhenxu94) [inline plugins/fallbacker/timer/timer_fallbacker.go:65]: ??

```suggestion
				return false
```

Comment (kezhenxu94) [reply plugins/fallbacker/timer/timer_fallbacker.go:65]: Add this test case please

**Review by kezhenxu94 — COMMENTED** [review]
Does the max attempts includes the original one (when the max attempts is set to 3 and it failed and comes to fallbacker, you only retry 3-1=2 times) ? If so ignore some of my comments. But others should be addressed.

**Comment by kezhenxu94** [comment]
> Does the max attempts includes the original one (when the max attempts is set to 3 and it failed and comes to fallbacker, you only retry 3-1=2 times) ? If so ignore some of my comments. But others should be addressed.

File: plugins/fallbacker/timer/timer_fallbacker.go (hunk 3 above)
Comment (EvanLjp) [reply plugins/fallbacker/timer/timer_fallbacker.go:60]: when trigger the fallbacker, already trigger once, so from 1 to add

**Pull request merged**